Supporting interface names in local-address and source

haverma opened the pull request:
- Supporting interface names in local-address of ISAKMP profiles and keyrings,  and in source of Tunnels
- Also added tests for ISAKMP profile

dhalperi commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkeGatewaylMatchers.java, line 14 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTN1GBzxReMPvHX08:-LDwTN1GBzxReMPvHX09:b-s221sw) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkeGatewaylMatchers.java#L14)):*
> ```
> l
> ```

typo in class name?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicylMatchers.java, line 7 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTTvbA0W3tjKMzKEY:-LDwTTvbA0W3tjKMzKEZ:blmivnw) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicylMatchers.java#L7)):*
> ```
> PolicylMatc
> ```

same typo? (extra `l`)

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicyMatchersImpl.java, line 12 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTbCa0CGTZjImPd1d:-LDwTbCa0CGTZjImPd1e:bwerwn8) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicyMatchersImpl.java#L12)):*
> ```
> pre-shared key hash:", "preSharedKeyHash
> ```

`pre-shared key hash` vs `preSharedKeyHash` -- should they be the same?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceAddressMatchersImpl.java, line 13 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTjOL37_RC99t-osP:-LDwTjOL37_RC99t-osQ:b33nmbr) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceAddressMatchersImpl.java#L13)):*
> ```
> IP:", "Ip
> ```

IP vs Ip?

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceMatchersImpl.java, line 43 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTpx_DqIRbt6N8Yo4:-LDwTpx_DqIRbt6N8Yo5:b-qsyuya) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceMatchersImpl.java#L43)):*
> ```Java
>   }
> 
>   static class HasAddress extends FeatureMatcher<Interface, InterfaceAddress> {
> ```

final?

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 5078 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwUL1S6a6ehOzOG7I0:-LDwUL1S6a6ehOzOG7I1:b-s4zbyd) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L5078)):*
> ```
>         _configuration.referenceStructure(INTERFACE, sourceInterfaceName, TUNNEL_SOURCE, line);
> ```

This should be moved out of the for loop -- I think you can put it in the `else` above. Then you can delete `line` and move it entirely into the `else` as well.

(Otherwise, we're referencing the interface once per `_currentInterfaces` rather than just once)

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 3972 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwYNTO1doQuOqTE6Ns:-LDwYNTO1doQuOqTE6Nt:bn343i5) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L3972)):*
> ```
> // LocalAddress value will be Ip.Auto(-1) if it contained an invalid interface name
> ```

Perhaps more clear:

`// LocalAddress can only be Ip.AUTO if LocalInterfaceName contains an invalid interface`

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 3974 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwXh_H8O6Rj-PyYgv0:-LDwXh_H8O6Rj-PyYgv1:b28fqst) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L3974)):*
> ```
>         if (isakmpProfile.getLocalAddress() != null
>             && isakmpProfile.getLocalAddress().equals(Ip.AUTO)) {
> ```

simpler: either `Objects.equals(isakmpProfile.getLocalAddress(), Ip.AUTO)` or `Ip.AUTO.equals(...)`

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4219 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwWoFq14_RBcaHUm_y:-LDwWoFq14_RBcaHUm_z:b-2p0pms) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4219)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
>               Ip localAddress = ifcNameIpAddress.get(keyring.getLocalInterfaceName());
>               if (localAddress != null) {
>                 keyring.setLocalAddress(localAddress);
>               } else {
>                 keyring.setLocalAddress(Ip.AUTO);
>               }
> ```
</details>


simpler:

```java
keyring.setLocalAddress(firstNonNull(ifcNameIpAddress.get(keyring.getLocalInterfaceName()), Ip.AUTO));
```

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4237 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwWWNz1q-IylLQs00l:-LDwWWNz1q-IylLQs00m:bv5ekia) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4237)):*
> ```Java
>   }
> 
>   private void resolveTunnelSourceInterfaces() {
> ```

document what this does.

looks like it's actually resolving addresses, not interfaces.

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4238 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwW6dP4n5Q9JBtIpQ0:-LDwW6dP4n5Q9JBtIpQ1:by5d7eb) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4238)):*
> ```Java
> 
>   private void resolveTunnelSourceInterfaces() {
>     Map<String, Ip> ifcNameIpAddress =
> ```

look at CommonUtil to see if a similar function exists and/or can be reused.

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 1627 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwVOPrAG7T8Lsy1jUv:-LDwVOPrAG7T8Lsy1jUw:b-ja5ulf) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L1627)):*
> ```Java
>             allOf(
>                 hasAddress(new Ip("1.2.3.4")),
>                 hasExternalInterface(InterfaceMatchers.hasAddress(hasIp(new Ip("2.3.4.6")))),
> ```

Per our in-person discussion, I think you can just use hasName using the super.getName() call.

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1549)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 8 of 16 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicyMatchersImpl.java, line 12 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTbCa0CGTZjImPd1d:-LDxCV8h2TiPddsnnb7x:b499q7j) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicyMatchersImpl.java#L12)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> pre-shared key hash:", "preSharedKeyHash
> ```

`pre-shared key hash` vs `preSharedKeyHash` -- should they be the same?
</blockquote></details>

Changed

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceMatchersImpl.java, line 43 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTpx_DqIRbt6N8Yo4:-LDxCb69ELmJJPNSSrTo:b3xqjyq) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceMatchersImpl.java#L43)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

final?
</blockquote></details>

Removed the `HasAddress` matcher as I am now using `HasName`

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 5078 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwUL1S6a6ehOzOG7I0:-LDxCtWYDTa-uIYt2kFy:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L5078)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
>         _configuration.referenceStructure(INTERFACE, sourceInterfaceName, TUNNEL_SOURCE, line);
> ```

This should be moved out of the for loop -- I think you can put it in the `else` above. Then you can delete `line` and move it entirely into the `else` as well.

(Otherwise, we're referencing the interface once per `_currentInterfaces` rather than just once)
</blockquote></details>

done

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 3972 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwYNTO1doQuOqTE6Ns:-LDxCvNhDnpE53DopQCa:b-n9tkhd) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L3972)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> // LocalAddress value will be Ip.Auto(-1) if it contained an invalid interface name
> ```

Perhaps more clear:

`// LocalAddress can only be Ip.AUTO if LocalInterfaceName contains an invalid interface`
</blockquote></details>

changed

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 3974 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwXh_H8O6Rj-PyYgv0:-LDxCypBBLv97fyL4Ut6:b-n9tkhd) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L3974)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
>         if (isakmpProfile.getLocalAddress() != null
>             && isakmpProfile.getLocalAddress().equals(Ip.AUTO)) {
> ```

simpler: either `Objects.equals(isakmpProfile.getLocalAddress(), Ip.AUTO)` or `Ip.AUTO.equals(...)`
</blockquote></details>

changed

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4219 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwWoFq14_RBcaHUm_y:-LDxD-kO5sAqGWLkOBhh:b-n9tkhd) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4219)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
>               Ip localAddress = ifcNameIpAddress.get(keyring.getLocalInterfaceName());
>               if (localAddress != null) {
>                 keyring.setLocalAddress(localAddress);
>               } else {
>                 keyring.setLocalAddress(Ip.AUTO);
>               }
> ```

simpler:

```java
keyring.setLocalAddress(firstNonNull(ifcNameIpAddress.get(keyring.getLocalInterfaceName()), Ip.AUTO));
```
</blockquote></details>

changed

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4237 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwWWNz1q-IylLQs00l:-LDxD1Fh8aUvjVjHsrbo:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4237)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

document what this does.

looks like it's actually resolving addresses, not interfaces.
</blockquote></details>

done

---

*[projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java, line 4238 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwW6dP4n5Q9JBtIpQ0:-LDxD7sd4gVn_ZJIGfOd:b-4nfupk) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/main/java/org/batfish/representation/cisco/CiscoConfiguration.java#L4238)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

look at CommonUtil to see if a similar function exists and/or can be reused.
</blockquote></details>

I didn't find any function which I can use here,  moved the logic out in a `private static` local helper

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 1627 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwVOPrAG7T8Lsy1jUv:-LDxDJl97Q27EazDkZzv:b-24fbsg) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L1627)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Per our in-person discussion, I think you can just use hasName using the super.getName() call.
</blockquote></details>

using `hasName` now

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkeGatewaylMatchers.java, line 14 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTN1GBzxReMPvHX08:-LDxCOwU1FfxaIOzLXhd:b-1dcph8) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkeGatewaylMatchers.java#L14)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> l
> ```

typo in class name?
</blockquote></details>

Corrected

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicylMatchers.java, line 7 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTTvbA0W3tjKMzKEY:-LDxCRT68tHmSyjfc6yB:b-rctorc) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/IkePolicylMatchers.java#L7)):*
> Corrected

---

*[projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceAddressMatchersImpl.java, line 13 at r1](https://reviewable.io/reviews/batfish/batfish/1549#-LDwTjOL37_RC99t-osP:-LDxDO4-CcKr2fesr6tQ:by4r21m) ([raw file](https://github.com/batfish/batfish/blob/06ec21d51b33284292b7cc172f2ed7af99548e48/projects/batfish-common-protocol/src/test/java/org/batfish/datamodel/matchers/InterfaceAddressMatchersImpl.java#L13)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

> ```
> IP:", "Ip
> ```

IP vs Ip?
</blockquote></details>

File deleted

---


*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1549)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/batfish/batfish/1549)*
<!-- Sent from Reviewable.io -->
